craft_store/request_client.py

```diff
@@ -0,0 +1,257 @@
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Client for making requests towards publisher gateway."""
+
+import contextlib
+from collections.abc import Generator
+from enum import Enum
+from json import JSONDecodeError
+from logging import getLogger
+from typing import Any, Literal
+
+import httpx
+
+from craft_store import auth, creds, errors
+
+logger = getLogger(__name__)
+
+
+class DeveloperTokenAuth(httpx.Auth):
+    """Request authentication using developer token."""
+
+    def __init__(
+        self,
+        *,
+        auth: auth.Auth,
+        auth_type: Literal["bearer", "macaroon"] = "bearer",
+    ) -> None:
+        self._auth = auth
+        self._auth_type = auth_type
+
+    def auth_flow(
+        self,
+        request: httpx.Request,
+    ) -> Generator[httpx.Request, httpx.Response, None]:
+        """Update request to include Authorization header."""
+        logger.debug("Adding developer token to authorize request")
+        if self._developer_token is None:
+            self.get_token_from_keyring()
+
+        self._update_headers(request)
+        yield request
+
+    def get_token_from_keyring(self) -> None:
+        """Get token stored in the credentials storage."""
+        logger.debug("Getting developer token from credential storage")
+        dev_token = creds.DeveloperToken.from_json_string(self._auth.get_credentials())
+        self._developer_token = dev_token.macaroon
+
+    def _update_headers(self, request: httpx.Request) -> None:
+        """Add token to the request."""
+        logger.debug("Adding ephemeral token to request headers")
+        if self._developer_token is None:
+            raise errors.DeveloperTokenUnavailableError(
+                message="Developer token is not available"
+            )
+        request.headers["Authorization"] = self._format_auth_header()
+
+    def _format_auth_header(self) -> str:
+        if self._auth_type == "bearer":
+            return f"Bearer {self._developer_token}"
+        return f"Macaroon {self._developer_token}"
+
+
+class StoreOperationStatus(str, Enum):
+    """Class that reflects status of ongoing task."""
+
+    SUCCEEDED = "succeeded"
+    FAILED = "failed"
+    RUNNING = "running"
+    WAITING = "waiting"
+
+    def is_ongoing(self) -> bool:
+        """Check if task is still running."""
+        return self in (
+            StoreOperationStatus.WAITING,
+            StoreOperationStatus.RUNNING,
+        )
```

Comment:
Should we include it all or should it be a part of Service that utilizes this client ?